<a href="https://colab.research.google.com/github/paascorb/MergeSortPython/blob/main/TrabajoEloy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Trabajo Mergesort Python

En este apartado vamos a realizar la parte obligatoria del trabajo, donde vamos a programar el método de ordenación Mergesort en Python dse manera secuencial, paralela y usando Numba.

## Utilidades

Primero, antes de ponernos a programar los algoritmos, necesitamos programar los métodos auxiliares.

In [6]:
from random import *

def generar_datos(tamaño):
  '''Genera 'tamaño' datos aleatorios dcon valores entre -1000 y 1000
     Siempre genera la misma secuencia de valores'''

  seed(22) # Fijamos la semilla
  return [randint(-1000, 1000) for i in range(0, tamaño)]

Probamos la primera función.

In [8]:
generar_datos(3)

[962, 858, -713]

Seguimos con las siguientes.

In [10]:
def esta_ordenado_array(vector, tamaño, ti, tf):
  '''Comprueba si un array está ordenado
     Devolverá la primera posición que no está ordenada y -1 si el vector está ordenado'''

  aux = -1

  for i in range(0, tamaño-1):
    if(vector[i]>vector[i+1]):
      aux = i
      i = tamaño

  if(aux == -1):

    print("El array está ordenado\n\n")
    print("Tiempo invertido en ordenar: "+str(tf-ti)+" milisegundos \n\n")

  else:
    print("\nEl vector falla en ****  "+ str(aux) +"  ******")

  return aux

In [11]:
esta_ordenado_array([1,2,3,4], 4, 100, 102)

El array está ordenado


Tiempo invertido en ordenar: 2 milisegundos 




-1

In [12]:
esta_ordenado_array([1,5,3,4], 4, 100, 102)


El vector falla en ****  1  ******


1

In [48]:
def burbuja(v, izq, der):
  '''Ordenación secuencial por el método de la burbuja
     Se pasan los índices izq y ordena desde v[iz] hasta v[der-1]'''
  aux = 0
  for i in range(0, der-izq-1):
    for j in range(izq, der-i-1):
      if v[j] > v[j+1]:
        aux = v[j]
        v[j]=v[j+1]
        v[j+1]=aux

In [49]:
prueba = [3,2,1]
burbuja(prueba,0,3)
prueba

[1, 2, 3]

In [50]:
import time

def milisegundos():
  # Devuelve el tiempo en milisengundos desde la época Unix (01/01/1970)

	return round(time.time() * 1000)

In [51]:
milisegundos()

1686675295683

In [55]:
def mezclar(v , izq, med, der):
  '''Mezcla ordenada de dos trozos de un vector desde sus índices izq hasta der
     Primer trozo desde izq hasta med-1
     Segundo trozo desde med hasta der-1
     Cada trozo tiene que estar previamente ordenado'''

  aux= [0] * (der-izq)
  i=izq
  j=med
  k=0

  while i<med and j<der:
    if v[i] <= v[j]:
      aux[k] = v[i]
      i += 1
    else:
      aux[k] = v[j];
      j += 1
    k += 1

  while i<med:
    aux[k] = v[i]
    k += 1
    i += 1

  while j<der:
    aux[k] = v[j];
    k += 1
    j += 1

  for i in range(0, der-izq):
      v[i+izq] = aux[i]

## Python Secuencial

En esta sección vamos a programar el algoritmo Mergesort secuencialmente.

In [59]:
def mergesort_secuencial(v, n, num_trozos):

  tam_trozo = int(n / num_trozos)
  for i in range(0, num_trozos):
    izq = i*tam_trozo
    der = (i+1) * tam_trozo
    if i == num_trozos-1: der = n
    burbuja(v, izq, der)

	# Mezclar todos los trozos ordenados
  for i in range(num_trozos-2, -1, -1):
    izq=i*tam_trozo
    med=izq+tam_trozo
    mezclar(v,izq,med,n)

Probamos el algoritmo y medimos tiempos.

In [74]:
def ejercicio1(n=None, num_trozos=None):
  print("ORDENACIÓN MERGESORT SECUENCIAL\n")

  if not n or not num_trozos:
    n = int(input("Escriba el tamaño del vector: ")) # Número de datos que se quieren ordenar
    num_trozos = int(input("Escriba cúantos trozos quiere usar: ")) # Número de trozos que se quieren usar

  # Generación aleatoria de n datos
  datos = generar_datos(n)
  # Ordenar
  print("Ordenando...")
  ti=milisegundos() # tiempo inicial
  mergesort_secuencial(datos,n,num_trozos)
  tf=milisegundos() # tiempo final

  # Comprobar que está correcto
  res = esta_ordenado_array(datos, n, ti, tf)
  if res == -1:
    return tf-ti

In [67]:
ejercicio1()

ORDENACIÓN MERGESORT SECUENCIAL

Escriba el tamaño del vector: 10
Escriba cúantos trozos quiere usar: 2
Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 0 milisegundos 




0

Teniendo esto ya, lo mejor es que calculemos los tiempos con los mismos datos que utilizábamos en la primera parte del trabajo. Recordemos que lo haciamos para tamaños de vectores 10, 1000, 10 000 y 1 000 000; y para trozos de tamaño 10, 100, 500 y 1000.

In [72]:
n_tro = [10, 100, 500, 1000]
tam = [10, 100, 10000, 1000000]

resultado = []
for n in tam:
  fila = []
  for t in n_tro:
    if n == 1000000 and t in [10, 100, 500]:
      fila.append(">100 000")
    else:
      fila.append(ejercicio1(n, t))

  resultado.append(fila)

ORDENACIÓN MERGESORT SECUENCIAL

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 0 milisegundos 


ORDENACIÓN MERGESORT SECUENCIAL

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 0 milisegundos 


ORDENACIÓN MERGESORT SECUENCIAL

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 2 milisegundos 


ORDENACIÓN MERGESORT SECUENCIAL

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 3 milisegundos 


ORDENACIÓN MERGESORT SECUENCIAL

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 0 milisegundos 


ORDENACIÓN MERGESORT SECUENCIAL

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 1 milisegundos 


ORDENACIÓN MERGESORT SECUENCIAL

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 12 milisegundos 


ORDENACIÓN MERGESORT SECUENCIAL

Ordenando...
El array está ordenado


Tiempo invertido en ordenar: 19 milisegundos 


ORDENACIÓN MERGESORT SECUENCIAL

Ordenando...
El array

In [83]:
resultado = [ [abs(i) for i in j if isinstance(i, int)] for j in resultado]

In [84]:
resultado

[[0, 0, 2, 3], [0, 1, 12, 19], [894, 218, 724, 1354], [387758]]

Los resultado de un millón no aparecen porque los he saltado, ya que sobrepasaban demasiado el tiempo de cálculo. Pero tenemos la tabla.

Tabla | 10 | 100 | 500 | 1000
--- | --- | --- | --- | ---
10| 0 | 0 | 2 | 3
1000 | 0 | 1 | 12 | 19
10 000 | 894 | 218 | 724 | 1354
1 000 000 | >100 000 | >100 000 | >100 000 | 358 758

## Python paralelo